src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,347 @@
+        }));
+      }
+  ));
+}
+
+// The capability that lets us call remote methods over RPC.
+// The client capability is dropped after each callRequest().
+class JsRpcTargetImpl final : public rpc::JsRpcTarget::Server {
+public:
+  JsRpcTargetImpl(
+      kj::Own<kj::PromiseFulfiller<void>> callFulfiller,
+      IoContext& ctx,
+      kj::Maybe<kj::StringPtr> entrypointName)
+      : callFulfiller(kj::mv(callFulfiller)), ctx(ctx), entrypointName(entrypointName) {}
+
+  // Handles the delivery of JS RPC method calls.
+  kj::Promise<void> call(CallContext callContext) override {
+    auto methodName = kj::heapString(callContext.getParams().getMethodName());
+    auto serializedArgs = callContext.getParams().getSerializedArgs().getV8Serialized().asBytes();
+
+    // Try to execute the requested method.
+    return ctx.run(
+        [this, methodName=kj::mv(methodName), serializedArgs=kj::mv(serializedArgs),
+          entrypointName=entrypointName, callContext] (Worker::Lock& lock) mutable ->
+            kj::Promise<void> {
+
+      jsg::Lock& js = lock;
+      if (!FeatureFlags::get(js).getJsRpc()) {
+        // JS RPC is not enabled on the server side, we cannot call any methods.
+        auto rpcDisabled =
+            "The receiving Worker does not allow its methods to be called over RPC."_kjc;
+        JSG_FAIL_REQUIRE(Error, rpcDisabled);
+      }
+
+      // We will try to get the function, if we can't we'll throw an error to the client.
+      auto fn = tryGetFn(lock, entrypointName, ctx, methodName);
+
+      // We have a function, so let's call it and serialize the result for RPC.
+      // We account for both the JS method throwing and our serialization attempt failing.
+      // If the function returns a promise we will wait for the promise to finish so we can
+      // serialize the result.
+      // TODO(now): Do we need to use awaitJs here? `call()` has to return a kj::Promise, so maybe
+      // we can return some other promise that resolves when this jsg::Promise does?
+      return ctx.awaitJs(js, callFuncAndSerialize(js, fn, serializedArgs).then(js,
+          [callContext](jsg::Lock& js, SerializedResult result) mutable {
+        KJ_SWITCH_ONEOF(result) {
+          KJ_CASE_ONEOF(ok, Returned) {
+            auto builder = callContext.initResults(capnp::MessageSize {
+                ok.serializedValue.size() / 8 + 8, 0 });
+            builder.initValue().setV8Serialized(kj::mv(ok.serializedValue));
+          }
+          KJ_CASE_ONEOF(err, ThrewException) {
+            auto builder = callContext.initResults(capnp::MessageSize {
+                err.serializedException.size() / 8 + 8, 0 });
+            builder.initException().setV8Serialized(kj::mv(err.serializedException));
+          }
         }
-    ));
+      }));
+    }).then([callFulfiller = kj::mv(callFulfiller)]() mutable {
+      // We want to fulfill the callPromise so customEvent can continue executing.
+      callFulfiller->fulfill();
```

Comment:
Previously you were also fulfilling `callFulfiller` here in the case of an exception, but now you aren't. So now if `run()` ends up throwing, `callFullfiller` will simply be dropped, which causes it to reject with a generic "PromiseFulfiller was never fulfilled" exception, which is not a friendly error. But I think you actually don't even want to reject `callFulfiller` in the case that the call throws, right? Because either way, the call is done now?

And what about if the call is canceled? I think that still counts as "done".

It seems like what you really want here is:

* Change `call()` back to a coroutine.
* `co_await` the result of `ctx.run()`.
* _Before_ the call to `run()`, use KJ_DEFER(callFulfiller->fulfill())`, so that the fulfiller is fulfilled when the call stops executing regardless of the outcome.

Reply:
Thanks, I've gone through each of the comments and applied the suggestions https://github.com/cloudflare/workerd/compare/a9254ed192e88344b38851274cb677fb36bd0fd6..9472684e8df98603e500ce62fe5021ba05cc50a5.

I also inlined a method (`verifyCallable`) because I realized the logic was a bit confusing and that lead to a subtle bug.